### PR TITLE
TEP-0142: Introduce StepAction referencing syntax in Steps

### PR DESCRIPTION
**Pull request by chitrangpatel**
This PR adds the referencing syntax for `StepActions` in `Stpes`. It also adds the necessary conversion between `v1beta1` and `v1` and the necessary validation. The actual resolution of the referenced `StepAction` will be implemented in a followup [PR](https://github.com/tektoncd/pipeline/pull/7271).

This is the third item on the implementation Issue https://github.com/tektoncd/pipeline/issues/7259

# Changes

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
[WIP] Introduces referencing syntax for StepAction in Steps
```
/kind feature

### CODE REVIEW

File: docs/stepactions.md
Comment (QuanZhang-William) [inline docs/stepactions.md:55]: double : ? (outside of this PR)

File: pkg/apis/pipeline/v1/task_validation_test.go
Comment (QuanZhang-William) [inline pkg/apis/pipeline/v1/task_validation_test.go:763]: I'm not sure if this test is really necessary? Seems like the code is already covered by the `TestTaskiSpecValidateErrorWithStepActionRef` test below, and this looks like a duplicate

File: pkg/apis/pipeline/v1beta1/container_types.go
Comment (QuanZhang-William) [inline pkg/apis/pipeline/v1beta1/container_types.go:232]: nit: StepAction

**Review by QuanZhang-William — COMMENTED** [review]
Thanks Chitrang, left some small comments

File: docs/stepactions.md
Comment (chitrangpatel) [reply docs/stepactions.md:75]: Done!

Comment (chitrangpatel) [reply docs/stepactions.md:75]: PTAL!

File: docs/stepactions.md
Comment (chitrangpatel) [reply docs/stepactions.md:55]: That's a typo. Thanks!

File: pkg/apis/pipeline/v1beta1/container_types.go
Comment (chitrangpatel) [reply pkg/apis/pipeline/v1beta1/container_types.go:232]: fixed

File: docs/stepactions.md
Comment (chitrangpatel) [reply docs/stepactions.md:55]: fixed

File: pkg/apis/pipeline/v1/task_validation_test.go
Comment (chitrangpatel) [reply pkg/apis/pipeline/v1/task_validation_test.go:763]: good point. I will remove it.

Comment (chitrangpatel) [reply pkg/apis/pipeline/v1/task_validation_test.go]: done.

**Comment by chitrangpatel** [comment]
> Thanks Chitrang, left some small comments

Thanks Quan! PTAL, I implemented them.

**Review by jerop — APPROVED** [review]

**Review by QuanZhang-William — APPROVED** [review]

**Comment by jerop** [comment]
@tektoncd/core-maintainers please take a look 🙏🏾

File: pkg/apis/pipeline/v1/task_validation_test.go
Comment (Yongxuanzhang) [inline pkg/apis/pipeline/v1/task_validation_test.go:1406]: nit: is there an extra `i` in the name?

Comment (chitrangpatel) [reply pkg/apis/pipeline/v1/task_validation_test.go]: Yes. Good eye 👁️ 

Fixed.

File: test/parse/yaml.go
Comment (chitrangpatel) [reply test/parse/yaml.go]: Good catch. I think it will be used in the next PR. I split up the large one into smaller ones so that's how it must have slipped in.

I removed it from here.

Thanks!

**Review by Yongxuanzhang — APPROVED** [review]

**Comment by jerop** [comment]
thank you for breaking down this work @chitrangpatel 🙏🏾 

/lgtm

---
adding this note as suggested in https://github.com/tektoncd/community/pull/1097#discussion_r1373208182:

sought more input here and on slack but haven't received any yet, we already have 3 approvals and this is specific change was discussed in a TEP plus it can be easily changed if needed (alpha), so merging to unblock progress for this series of pull requests targeting v0.54

**Pull request merged**
